xmr-btc/src/alice.rs

```diff
@@ -1,31 +1,167 @@
+    rng: &mut R,
+) -> Result<State> {
+    match state {
+        State::State0(state0) => {
+            transport
+                .send_message(state0.next_message(rng).into())
+                .await?;
 
-#[derive(Debug)]
-pub struct Message0 {
-    pub(crate) A: bitcoin::PublicKey,
-    pub(crate) S_a_monero: monero::PublicKey,
-    pub(crate) S_a_bitcoin: bitcoin::PublicKey,
-    pub(crate) dleq_proof_s_a: cross_curve_dleq::Proof,
-    pub(crate) v_a: monero::PrivateViewKey,
-    pub(crate) redeem_address: bitcoin::Address,
-    pub(crate) punish_address: bitcoin::Address,
+            let bob_message0: bob::Message0 = transport.receive_message().await?.try_into()?;
+            let state1 = state0.receive(bob_message0)?;
+            Ok(state1.into())
+        }
+        State::State1(state1) => {
+            let bob_message1: bob::Message1 = transport.receive_message().await?.try_into()?;
+            let state2 = state1.receive(bob_message1);
+            let alice_message1: Message1 = state2.next_message();
+            transport.send_message(alice_message1.into()).await?;
+            Ok(state2.into())
+        }
+        State::State2(state2) => {
+            let bob_message2: bob::Message2 = transport.receive_message().await?.try_into()?;
+            let state3 = state2.receive(bob_message2)?;
+            Ok(state3.into())
+        }
+        State::State3(state3) => {
+            tracing::info!("alice is watching for locked btc");
+            let state4 = state3.watch_for_lock_btc(bitcoin_wallet).await?;
+            Ok(state4.into())
+        }
+        State::State4(state4) => {
+            let state5 = state4.lock_xmr(monero_wallet).await?;
+            tracing::info!("alice has locked xmr");
+            Ok(state5.into())
+        }
+        State::State5(state5) => {
+            transport.send_message(state5.next_message().into()).await?;
+            // todo: pass in state4b as a parameter somewhere in this call to prevent the
+            // user from waiting for a message that wont be sent
+            let message3: bob::Message3 = transport.receive_message().await?.try_into()?;
+            let state6 = state5.receive(message3);
+            tracing::info!("alice has received bob message 3");
+            tracing::info!("alice is redeeming btc");
+            state6.redeem_btc(bitcoin_wallet).await?;
+            Ok(state6.into())
+        }
+        State::State6(state6) => Ok(state6.into()),
+    }
 }
 
+#[allow(clippy::large_enum_variant)]
 #[derive(Debug)]
-pub struct Message1 {
-    pub(crate) tx_cancel_sig: Signature,
-    pub(crate) tx_refund_encsig: EncryptedSignature,
+pub enum State {
+    State0(State0),
+    State1(State1),
+    State2(State2),
+    State3(State3),
+    State4(State4),
+    State5(State5),
+    State6(State6),
 }
 
-#[derive(Debug)]
-pub struct Message2 {
-    pub(crate) tx_lock_proof: monero::TransferProof,
+// TODO: use macro or generics
+pub fn is_state4(state: &State) -> bool {
```

Comment:
As far as I could see, these are only used for the test right? Could move it to the harness test harness then and keep the library slim.